scalyr_agent/log_processing.py

```diff
@@ -2282,6 +2322,46 @@ def config(self):
         """
         return self.__log_entry_config
 
+    def finish( self ):
+        """
+        Tells the log matcher to finish processing any processors.
+
+        LogMatchers that are `finished` no longer match log files or create new processors, unless
+        they haven't ever matched any files, in which case they will match at least once
+        """
+        self.__lock.acquire()
+        try:
+            # set the matcher as finished
+            self.__is_finished = True
+
+            # set any existing processors as finished
+            for processor in self.__processors:
+                processor.finish()
+        finally:
+            self.__lock.release()
+
+    def is_finished( self ):
+        """
+        Returns true if the log matcher and all of its processors are finished, and the matcher has
+        been processed at least once
+        """
+        result = False
+        self.__lock.acquire()
+        try:
+            # check if we are finished
+            result = self.__is_finished and self.__last_check is not None
+
+            # check if all the processors are finished
+            for processor in self.__processors:
+                # exit early if the matcher or a previous processor is finished
+                if not result:
```

Comment:
If you do the early exit above, this one can be written in a more obvious manner:
```
 if not processor.is_finished:
     return False
```